monome/teletype
# Allow calibration of CV outputs with CV.CAL and CV.CAL.RESET

#### What does this PR do?

Adds a CV.CAL op to create calibration scaling parameters for a CV output and store them in flash memory, and a CV.CAL.RESET op to restore the scaling to the default.

Using the op to calibrate requires a digital multimeter with at least millivolt precision. The procedure is:

1. run `L 1 4: CV I V 1` and record measurements to 3 decimal places with a good digital multimeter.
2. run `L 1 4: CV I V 3` and record measurements to 3 decimal places etc. (second column)
3. run `CV.CAL <output num> <V 1 measurement in millivolts> <V 3 measurement in millivolts>` for each output.

For example, I measure the following on my unit:

![](https://media.discordapp.net/attachments/850955610471661610/1106039733093679164/image.png)

So with that “before” data I enter:
```
CV.CAL 1 989 2982
CV.CAL 2 1017 3011
CV.CAL 3 1002 2997
CV.CAL 4 996 2984
```

and afterwards I observe the values reported in the “after” columns above.

#### Provide links to any related discussion on [lines](https://llllllll.co/).

https://llllllll.co/t/teletype-cv-calibration/5654

#### How should this be manually tested?

Follow the calibration procedure described above and measure the before and after states. Power off the module and turn it back on and Re measure to confirm that the calibration was saved and is still being applied. 

#### Any background context you want to provide?

For a couple years now I have been using a patch to do calibration with floating-point math and hardcoded constants (as described in the lines thread), and I finally got around to rewriting it using fixed point integer math and implementing an op for it so I could more easily use the 5.0 beta branch for an upcoming performance.

#### If the related Github issues aren't referenced in your commits, please link to them here.

#145 

#### I have,
